### PR TITLE
[sharktank] Refactor mask generation to utils from ops

### PR DESCRIPTION
**Pull request by archana-ramalingam**
Revert the mask generation functions from ops back to utils

### CODE REVIEW

File: sharktank/sharktank/utils/attention.py [hunk 1]
```diff
@@ -75,3 +143,75 @@ def create_boolean_chunked_attention_mask(
...
+
+    assert (
+        batch_seq_len % attention_chunk_size == 0
+    ), f"Sequence length ({batch_seq_len}) must be divisible by attention chunk size ({attention_chunk_size})"
+
+    attention_mask = unbox_tensor(attention_mask)
+
+    device = attention_mask.device
+
+    # TODO: handle decode step
```
[the remark this answers is not in the document]
Comment (Alex-Vasile) [reply sharktank/sharktank/utils/attention.py]: I would actually leave the TODOs, but tag them to an issue. This PR is likely going to land before either of those two.

File: sharktank/sharktank/utils/attention.py [hunk 2]
```diff
@@ -75,3 +143,75 @@ def create_boolean_chunked_attention_mask(
...
+        chunked_boolean_attention_mask,
+        attention_mask,
+        torch.tensor(
+            max_negative_value(attention_mask.dtype, device=device),
+            dtype=attention_mask.dtype,
+        ),
+    )
+
+
+def create_input_mask(seq_lens: torch.Tensor, batch_seqlen: int) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (archana-ramalingam) [reply sharktank/sharktank/utils/attention.py]: Added `trivially_replicable` and further fix by Boian here https://github.com/nod-ai/shark-ai/pull/2522, to allow it to be used for functions.

File: sharktank/sharktank/utils/attention.py [hunk 3]
```diff
@@ -6,16 +6,68 @@
...
+    boolean_input_mask = unbox_tensor(boolean_input_mask)
+    device = boolean_input_mask.device
+
+    dtype = (
+        torch.float32 if attention_dtype == torch.float8_e4m3fnuz else attention_dtype
+    )
+    numeric_mask = torch.where(
+        boolean_input_mask, max_negative_value(dtype, device), 0
+    ).to(dtype)
+    return numeric_mask.unsqueeze(1).unsqueeze(1).to(device)
```
Comment (Alex-Vasile) [inline sharktank/sharktank/utils/attention.py]: Is this `.to(device)` needed?

File: sharktank/sharktank/utils/attention.py [hunk 4]
```diff
@@ -6,16 +6,68 @@
...
+        torch.float32 if attention_dtype == torch.float8_e4m3fnuz else attention_dtype
+    )
+    causal_mask = create_causal_context_mask(
+        source_len=source_len,
+        target_len=target_len,
+        start_positions=start_positions,
+        device=device,
+    )
+    boolean_mask = torch.logical_or(causal_mask, boolean_input_mask[:, None, None, :])
+    numeric_mask = torch.where(boolean_mask, max_negative_value(dtype, device), 0).to(
```
Comment (sogartar) [inline sharktank/sharktank/utils/attention.py]: If we have `torch.logical_or` and `torch.where` implemented as `sharktank.ops`, the this function does not need to be decorated as trivially replicable. The same argument holds for some other functions in `create_attention_mask_for_decode` and `create_chunked_attention_mask`.
We generally want to expand the op converge to all that torch has. Definitely to all that we use.

I see that the problem is [create_causal_context_mask](https://github.com/nod-ai/shark-ai/blob/b65fdae35ede615932734d6c204771a33cfdc7ec/sharktank/sharktank/utils/attention.py#L17). Because it creates un-replicated tensors with
```
    src = torch.arange(src_len, device=device)[None, None, None, :]
    target = torch.arange(target_len, device=device)[None, None, :, None]
```
We will keep running into the problem of having to construct something that does not depend on a tensor arg, so there is nothing to propagate the sharding nature from. E.g. construct all-zeros or all-ones tensor. To generalize this approach we need to come up with something. Maybe allow downstream ops to mix sharded and unsharded args.

If we want to write generic model code we need a solution or we will keep dancing around the problem. This can make things quite nasty if you need to modify existing code to create such tensors where sharded tensors would be present. This can happen for example when extending our LLM to support some new architecture.

File: sharktank/tests/models/llama/attention_test.py [hunk 5]
```diff
@@ -49,15 +45,6 @@ def test(self, prefill_offset: bool):
             feature_dim=head_count * head_dim, ffn_dim=ffn_dim, dtype=torch.float32
         )
 
-        start_positions = torch.arange(0, bs)
```
[the remark this answers is not in the document]
Comment (archana-ramalingam) [reply sharktank/tests/models/llama/attention_test.py]: We have a separate eval test Rob added and Boian is working to add, this doesn't test the chunked prefill effectively. So removing it, but this needs to happen in a separate PR.

File: sharktank/sharktank/utils/attention.py [hunk 6]
```diff
@@ -35,16 +87,17 @@ def create_causal_context_mask(
...
+    source = torch.arange(source_len, device=device)[None, None, None, :]
     target = torch.arange(target_len, device=device)[None, None, :, None]
 
     if start_positions is not None:
         target = target + start_positions[:, None, None, None]
 
-    mask = src > target
-    return mask
+    mask = source > target
+    return mask.to(device)
```
[the remark this answers is not in the document]
Comment (archana-ramalingam) [reply sharktank/sharktank/utils/attention.py]: Can remove this, must have missed out during testing.

**Review by Alex-Vasile — APPROVED** [review]

File: sharktank/sharktank/utils/attention.py [hunk 7]
```diff
@@ -7,14 +7,66 @@
...
+    )
+    numeric_mask = torch.where(
+        boolean_input_mask, max_negative_value(dtype, device), 0
+    ).to(dtype)
+    return numeric_mask.unsqueeze(1).unsqueeze(1)
 
 
+@trivially_replicable
 def create_causal_context_mask(
     src_len: int,
```
Comment (Alex-Vasile) [inline sharktank/sharktank/utils/attention.py]:
```suggestion
    source_len: int,
```
